File: Shortest Path Using Atmost One Curved Edge.cpp

class Solution {
  public:
  
    vector<int> dijkstra(int u, int b, int n, vector<pair<int,int>> adj[]) {
        vector<int> dis;
        dis.assign(n+1, 1000000001);
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        dis[u] = 0;
        pq.push({0, u});
     
        while(!pq.empty()) {
            int u = pq.top().second;
            pq.pop();
    
            for(auto p: adj[u]) {
                  int v = p.first;
                  int w = p.second;
                  if(dis[v] > dis[u] + w) {
                        dis[v] = dis[u] + w;
                        pq.push({dis[v], v});
                  }
            }
        }
        return dis;
    }

    int shortestPath(int n, int m, int a, int b, vector<vector<int>> &edges) {
        vector<pair<int,int>> adj[n+1];
        vector<vector<int>> curved;
        
        for(int i=0; i<m; i++)
        {
            int u = edges[i][0];
            int v = edges[i][1];
            int w = edges[i][2];
            int cw = edges[i][3];
            
            adj[u].push_back({v, w});
            adj[v].push_back({u, w});
            
            // curved edge weight
            curved.push_back({u, v, cw});
        }
        
        vector<int> da = dijkstra(a, b, n, adj);
        vector<int> db = dijkstra(b, a, n, adj);
        
        int ans = da[b];
 
        // ans = min distance from a -> b with at max one curved edge 
        // current ans without curved edge from a -> b = da[b]
    
        for(int i=0; i<m; i++)
        {
            int u = curved[i][0];
            int v = curved[i][1];
            int cw = curved[i][2];
    
            ans = min(ans, da[u] + cw + db[v]);
            ans = min(ans, da[v] + cw + db[u]);
        }
        
        if(ans>=1000000001) return -1;
        return ans;
    }
};

File: Shortest Path Using Atmost One Curved Edge.py
class Solution:
    def dijkstra(self, u, b, n, adj):
        unused = [] 
        dis = [float('inf')]*(n+1)
        
        dis[u] = 0
        
        unused.append((0,u))
        
        heapify(unused)
        
        while unused:
            x, v = heappop(unused)
         
            if x > dis[v]:
                continue
            
            for edge in adj[v]:
                u = edge[0]
                w = edge[1]
                
                if dis[v] + w < dis[u]:
                    dis[u] = dis[v] + w
                    heappush(unused, (dis[u], u))
        
        return dis
        
    def shortestPath(self, n, m, a, b, edges):
        adj = [[] for i in range(n+1)]
        curved = []
        
        for i in range(m):
            u = edges[i][0]
            v = edges[i][1]
            w = edges[i][2]
            cw = edges[i][3]
            
            adj[u].append((v,w))
            adj[v].append((u,w))
            
            # curved edge weight
            curved.append([u, v, cw])
        
        da = self.dijkstra(a,b,n,adj)
        db = self.dijkstra(b,a,n,adj)
        
        ans = da[b]
        
        # ans = min distance from a -> b with at max one curved edge 
        # current ans without curved edge from a -> b = da[b]
    
        for i in range(m):
            u = curved[i][0]
            v = curved[i][1]
            cw = curved[i][2]
    
            ans = min(ans, da[u] + cw + db[v])
            ans = min(ans, da[v] + cw + db[u])
            
        if ans >= 1000000001:
            return -1
        return ans
